keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,858 @@
+| `host`        | string           | Host name to connect to.                                                                       |
+| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                          |
+| `path`        | string           | Path to access on the HTTP server.                                                             |
+| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. |
+| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                 |
+
+The `httpHeaders` field has type [`HTTPHeader array`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)
+with the following specification:
+
+(NOTE: We will need to evaluate how feasible it is to extend `HTTPHeader v1 core` with `valueFrom` to enable mounting headers from Secrets,
+for authentication. This may need to be its own independent KEP.)
+
+| Field       | Type             | Description                                                            |
+|-------------|------------------|------------------------------------------------------------------------|
+| `name`      | string           | The header field name                                                  |
+| `value`     | string           | The header field value                                                 |
+| `valueFrom` | HTTPHeaderSource | Source for the header's value. Cannot be used if `value` is non-empty. |
+
+The `valueFrom` field has type `HTTPHeaderSource` with the following specification:
+
+| Field          | Type              | Description                                             |
+|----------------|-------------------|---------------------------------------------------------|
+| `secretKeyRef` | SecretKeySelector | Selects a key of a secret in the ReplicaSet's namespace |
+
+The `secretKeyRef` field has type [`SecretKeySelector`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#secretkeyselector-v1-core):
+
+| Field  | Type   | Description                                                       |
+|--------|--------|-------------------------------------------------------------------|
+| `key`  | string | The key of the secret to select from. Must be a valid secret key. |
+| `name` | string | Name of the referent.                                             |
+| `key`  | string | Specify whether the Secret or its key must be defined.            |
+
+
+#### "Pod Picker" API contract:
+
+The contract for a `downscalePodPicker` API will be as follows:
+
+- Request payloads to the API will contain:
+   - `number_of_pods_requested` (`int`): minimum number of Pods to return
+   - `candidate_pods` (`list[string]`): list of Pod-names to choose from
```

Comment:
All lists need bounds.  Can I ask for 10K victim pods?

Reply:
It will be the Kubernetes controller performing the request, so hopefully `number_of_pods_requested` is never 10k pods!

However, naturally the `number_of_pods_requested` can not exceed the length of `candidate_pods`, as we are choosing from that list.

Reply:
Assume I have a 10K pod deployment and I set replicas to 10 (oops!) - Will this try to process 9990 victims?  As a spec it needs to be clear.

Reply:
While this use case would be unusual, I am not sure how the spec can be more clear on this?

If the `number_of_pods_requested` is `9990`, the Pod-Picker API should try to return as close to that many `chosen_pods` and `tied_pods` (from the `candidate_pods`) as possible.

However, I think one thing can be made more clear:
1. The kube-controller will not reject the Pod-Picker's response if it fails to return at least `number_of_pods_requested` Pods (that is, Pod-Picker APIs can be best-effort)